x/storage/keeper/keeper_rate_limit.go

```diff
@@ -0,0 +1,232 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	paymenttypes "github.com/bnb-chain/greenfield/x/payment/types"
+	"github.com/bnb-chain/greenfield/x/storage/types"
+	storagetypes "github.com/bnb-chain/greenfield/x/storage/types"
+)
+
+func (k Keeper) SetBucketFlowRateLimit(ctx sdk.Context, operator sdk.AccAddress, bucketOwner sdk.AccAddress, paymentAccount sdk.AccAddress, bucketName string, rateLimit sdkmath.Int) error {
+	// check the operator is the same as the payment account owner
+	if !k.paymentKeeper.IsPaymentAccountOwner(ctx, paymentAccount, operator) {
+		return paymenttypes.ErrNotPaymentAccountOwner
+	}
+
+	// get the bucket
+	bucket, found := k.GetBucketInfo(ctx, bucketName)
+	if !found {
+		return types.ErrNoSuchBucket
+	}
+
+	// check the bucket owner is the same as the bucket owner
+	if bucket.Owner != bucketOwner.String() {
+		return types.ErrInvalidBucketOwner
+	}
+
+	// if the bucket does not use the payment account, just set the flow rate limit
+	if bucket.PaymentAddress != paymentAccount.String() {
+		// set the flow rate limit to the store
+		k.setBucketFlowRateLimit(ctx, paymentAccount, bucketName, &types.BucketFlowRateLimit{
+			FlowRateLimit: rateLimit,
+		})
+		return nil
+	}
+
+	// below are the logic for the bucket using the payment account
+	if rateLimit.IsZero() {
+		err := k.setZeroBucketFlowRateLimit(ctx, bucket, paymentAccount, bucketName)
+		if err != nil {
+			return err
+		}
+	} else {
+		err := k.setNonZeroBucketFlowRateLimit(ctx, bucket, paymentAccount, bucketName, rateLimit)
+		if err != nil {
+			return err
+		}
+	}
+
+	// set the flow rate limit to the store
+	k.setBucketFlowRateLimit(ctx, paymentAccount, bucketName, &types.BucketFlowRateLimit{
```

Comment:
I think `bucketOwner` should be one of the key?  Otherwise someone can back run this tx, create a same `bucketName` before the expected create bucket tx come in.